kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt

```diff
@@ -138,23 +155,41 @@ internal class WorkQueue {
             } // Failed -> someone else stole it
         }
 
+        return pollWithMode(onlyBlocking = true /* only blocking */)
+    }
+
+    fun pollCpu(): Task? {
+        while (true) { // Poll the slot
+            val lastScheduled = lastScheduledTask.value ?: break
+            if (lastScheduled.isBlocking) break
+            if (lastScheduledTask.compareAndSet(lastScheduled, null)) {
+                return lastScheduled
+            } // Failed -> someone else stole it
+        }
+
+        return pollWithMode(onlyBlocking = false /* only cpu */)
+    }
+
+    private fun pollWithMode(/* Only blocking OR only CPU */ onlyBlocking: Boolean): Task? {
         val start = consumerIndex.value
         var end = producerIndex.value
-
-        while (start != end && blockingTasksInBuffer.value > 0) {
-            val task = tryExtractBlockingTask(--end)
+        // CPU or (BLOCKING & hasBlocking)
+        val shouldProceed = !onlyBlocking || blockingTasksInBuffer.value > 0
+        while (start != end && shouldProceed) {
+            val task = tryExtractFromTheMiddle(--end, onlyBlocking)
             if (task != null) {
                 return task
             }
         }
         return null
     }
 
-    private fun tryExtractBlockingTask(index: Int): Task? {
+    private fun tryExtractFromTheMiddle(index: Int, onlyBlocking: Boolean): Task? {
+        if (onlyBlocking && blockingTasksInBuffer.value == 0) return null
```

Comment:
I don't get it. Why not do this outside the function, in the caller? Then, we could leave the loop immediately (like it was done before the change), and `shouldProceed` wouldn't be needed.